### PR TITLE
Handle null cases of task input snapshots.

### PR DESCRIPTION
**Pull request by alkemist**
https://builds.gradle.org/project.html?projectId=Gradle_Branches&tab=projectOverview&branch_Gradle_Branches=ld%2Fbuild-ops%2Fsnapshot-inputs

### CODE REVIEW

**Comment by wolfs** [comment]
@alkemist What is the motivation for this change?

File: subprojects/core/src/main/java/org/gradle/caching/internal/tasks/BuildCacheKeyInputs.java [hunk 1]
```diff
@@ -37,7 +36,14 @@
     private final ImmutableSortedMap<String, HashCode> inputHashes;
     private final ImmutableSortedSet<String> outputPropertyNames;
 
-    public BuildCacheKeyInputs(String taskClass, HashCode classLoaderHash, List<HashCode> actionClassLoaderHashes, ImmutableList<String> actionClassNames, ImmutableSortedMap<String, HashCode> inputHashes, ImmutableSortedSet<String> outputPropertyNames) {
+    public BuildCacheKeyInputs(
+        @Nullable String taskClass,
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/caching/internal/tasks/BuildCacheKeyInputs.java]: I don't undertand why all these properties now need to be `@Nullable`. Especially all the lists should never be empty, I guess.

Comment (alkemist) [inline subprojects/core/src/main/java/org/gradle/caching/internal/tasks/BuildCacheKeyInputs.java]: This is used as a “null object” already, when a task has no outputs. Having unclear null semantics is a recipe for disaster, as evidenced by the fact that we did not realise that these objects can be null under some circumstances.

Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/caching/internal/tasks/BuildCacheKeyInputs.java]: I see how `classLoaderHash` can be null, but the other properties should never be null.

Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/caching/internal/tasks/BuildCacheKeyInputs.java]: So the problem is the construction of `NO_CACHE_KEY` here: https://github.com/gradle/gradle/blob/094a5ed82fcc9147de09afcc1970cb6260bba9f9/subprojects/core/src/main/java/org/gradle/caching/internal/tasks/DefaultTaskOutputCachingBuildCacheKeyBuilder.java#L38-L38?

`inputHashes` and `outputPropertyNames` cannot be null after all I saw.

**Comment by facewindu** [comment]
I merged it. The failing test is fixed, and the others are flakey tests.

**Comment by alkemist** [comment]
@wolfs I'm going to open up a PR for this area on master to make the null chain clearer.

**Pull request merged**